query/src/main/java/org/infinispan/query/dsl/embedded/impl/QueryEngine.java

```diff
@@ -734,113 +695,92 @@ public HsQueryRequest createHsQuery(String queryString, IndexedTypeMap<CustomTyp
-      org.apache.lucene.search.Query luceneQuery = makeTypeQuery(luceneParsingResult.getQuery(), luceneParsingResult.getTargetEntityName());
+      return transformParsingResult(parsingResult, namedParameters);
+   }
 
-      if (log.isDebugEnabled()) {
-         log.debugf("The resulting Lucene query is : %s", luceneQuery.toString());
-      }
+   <E> CacheQuery<E> buildLuceneQuery(IckleParsingResult<TypeMetadata> ickleParsingResult, Map<String, Object> namedParameters, long startOffset, int maxResults) {
+      return buildLuceneQuery(ickleParsingResult, namedParameters, startOffset, maxResults, IndexedQueryMode.FETCH);
+   }
 
-      if (queryMode == null) queryMode = detectQueryMode();
+   private IndexedQueryMode detectQueryMode() {
+      CacheMode cacheMode = cache.getCacheConfiguration().clustering().cacheMode();
+      boolean localCache = !cacheMode.isClustered();
+      boolean replicated = cacheMode.isReplicated();
+      return localCache || replicated ? IndexedQueryMode.FETCH : IndexedQueryMode.BROADCAST;
+   }
 
-      CacheQuery<?> cacheQuery = makeCacheQuery(ickleParsingResult, luceneQuery, queryMode, namedParameters);
-      // No need to set sort and projection if BROADCAST, as both are part of the query string already.
-      if (queryMode != IndexedQueryMode.BROADCAST) {
-         if (luceneParsingResult.getSort() != null) {
-            cacheQuery = cacheQuery.sort(luceneParsingResult.getSort());
-         }
-         if (luceneParsingResult.getProjections() != null) {
-            cacheQuery = cacheQuery.projection(luceneParsingResult.getProjections());
-         }
-      }
+   public <E> CacheQuery<E> buildLuceneQuery(IckleParsingResult<TypeMetadata> ickleParsingResult, Map<String, Object> namedParameters, long startOffset, int maxResults, IndexedQueryMode queryMode) {
+      CacheQuery<E> cacheQuery = (CacheQuery<E>) makeCacheQuery(ickleParsingResult, queryMode, namedParameters);
       if (startOffset >= 0) {
          cacheQuery = cacheQuery.firstResult((int) startOffset);
       }
       if (maxResults > 0) {
          cacheQuery = cacheQuery.maxResults(maxResults);
       }
-
-      return (CacheQuery<E>) cacheQuery;
+      return cacheQuery;
    }
 
-   private LuceneQueryParsingResult<TypeMetadata> transformParsingResult(IckleParsingResult<TypeMetadata> parsingResult, Map<String, Object> namedParameters) {
-      return queryCache != null && parsingResult.getParameterNames().isEmpty()
-            ? queryCache.get(parsingResult.getQueryString(), null, LuceneQueryParsingResult.class, (queryString, accumulators) -> transformToLuceneQueryParsingResult(parsingResult, namedParameters))
-            : transformToLuceneQueryParsingResult(parsingResult, namedParameters);
+   public SearchQuery<?> transformParsingResult(IckleParsingResult<TypeMetadata> parsingResult, Map<String, Object> namedParameters) {
+      // TODO HSEARCH-3896 Allow to reuse query instance
```

Comment:
You're right, this is not related to HSEARCH-3896. Because the issue is about reuse query instance with the same thread.
Yes. I'm going to restore something we did some time ago, that is basically Search 6 version of `LuceneQueryParsingResult`. Maybe it should help also to restore broadcast query.

Reply:
It seems that all cached queries here are stored in a special cache named `___query_cache`.

The scope of this cache is unfortunately **global**, which means that the instances here are supposed to be reused by different cache managers, that implies that the instances are supposed to be reused by different search mapping. :(

I think that with our current scope, projection, predicate and sort instances are supposed to be used only by the original search mapping, the one from which they have been created. Right?

I found this issue thanks to the test `org.infinispan.query.blackbox.IndexedCacheRestartTest` in which a cached query is created by a search mapping and used by another search mapping.
In this case, on the latter invokation the cached query parsing contains reference to closed service, e.g.: time service.

I’m trying now to not cache the scope… I don’t know actually if it make sense... Does it?

Reply:
The solution was [the one we discussed on Zulip](https://infinispan.zulipchat.com/#narrow/stream/118645-infinispan/topic/Query.20caching/near/195690950). It sounds very good! Adding it in a different commit.